File: e2e/playwright-typescript/interactions/steps/TaskTableStep.ts
import { expect, Page } from '@playwright/test';
import { step } from '@/decorators/step-decorator';
import { MainPage } from '@/interactions/pages/MainPage';
import { CreateTaskForm } from '@/interactions/pages/CreateTaskForm';
import { CreateTaskStep } from '@/interactions/steps/CreateTaskStep';
import { EditTaskStep } from '@/interactions/steps/EditTaskStep';
import { TaskInfoModal } from '@/interactions/pages/TaskInfoModal';
import { EditTaskForm } from '@/interactions/pages/EditTaskForm';

export class TaskTableStep {
  private readonly mainPage: MainPage;
  private readonly createTaskForm: CreateTaskForm;
  private readonly taskInfoModal: TaskInfoModal;
  private readonly editTaskForm: EditTaskForm;
  readonly createTask: CreateTaskStep;
  readonly editTask: EditTaskStep;

  constructor(page: Page) {
    this.mainPage = new MainPage(page);
    this.createTaskForm = new CreateTaskForm(page);
    this.taskInfoModal = new TaskInfoModal(page);
    this.editTaskForm = new EditTaskForm(page);
    this.createTask = new CreateTaskStep(page);
    this.editTask = new EditTaskStep(page);
  }

  @step('Open create task form')
  async openCreateTaskForm(): Promise<void> {
    await this.mainPage.createTaskButton.click();
    await expect(this.createTaskForm.submitButton).toBeVisible();
  }

  @step("Open task info for task '{title}'")
  async openTaskInfoForm(title: string): Promise<void> {
    const id = await this.resolveTaskId(title);
    await this.mainPage.taskInfoButtonById(id).click();
    await expect(this.taskInfoModal.title).toBeVisible();
  }

  @step("Open edit form for task '{title}'")
  async openEditTaskForm(title: string): Promise<void> {
    const id = await this.resolveTaskId(title);
    await this.mainPage.taskEditButtonById(id).click();
    await expect(this.editTaskForm.submitButton).toBeVisible();
  }

  @step("Delete task '{title}'")
  async deleteTask(title: string): Promise<void> {
    const id = await this.resolveTaskId(title);
    const deleteButton = this.mainPage.taskDeleteButtonById(id);
    await deleteButton.click();
    await expect(deleteButton).not.toBeVisible();
  }

  private async resolveTaskId(title: string): Promise<string> {
    const dataTestId = await this.mainPage.taskTitleByTitle(title).getAttribute('data-testid');
    return dataTestId!.replace('task-title-', '');
  }
}

File: e2e/playwright-python/interactions/steps/task_table_step.py
from playwright.sync_api import Page, expect
from interactions.pages.main_page import MainPage
from interactions.pages.create_task_form import CreateTaskForm
from interactions.pages.edit_task_form import EditTaskForm
from interactions.pages.task_info_modal import TaskInfoModal
from interactions.steps.create_task_step import CreateTaskStep
from interactions.steps.edit_task_step import EditTaskStep
from decorators.step_decorator import step


class TaskTableStep:
    def __init__(self, page: Page):
        self._page = page
        self._main_page = MainPage(page)
        self._create_task_form = CreateTaskForm(page)
        self._task_info_modal = TaskInfoModal(page)
        self._edit_task_form = EditTaskForm(page)

    @step("Open create task form")
    def open_create_task_form(self) -> CreateTaskStep:
        self._main_page.create_task_button.click()
        expect(self._create_task_form.submit_button).to_be_visible()
        return CreateTaskStep(self._page)

    @step("Open task info for task '{title}'")
    def open_task_info_form(self, title: str) -> None:
        task_id = self._resolve_task_id(title)
        self._main_page.task_info_button_by_id(task_id).click()
        expect(self._task_info_modal.title).to_be_visible()

    @step("Open edit form for task '{title}'")
    def open_edit_task_form(self, title: str) -> EditTaskStep:
        task_id = self._resolve_task_id(title)
        self._main_page.task_edit_button_by_id(task_id).click()
        expect(self._edit_task_form.title_input).to_have_value(title)
        return EditTaskStep(self._page)

    @step("Delete task '{title}'")
    def delete_task(self, title: str) -> None:
        task_id = self._resolve_task_id(title)
        delete_button = self._main_page.task_delete_button_by_id(task_id)
        delete_button.click()
        expect(delete_button).not_to_be_visible()

    def _resolve_task_id(self, title: str) -> str:
        data_test_id = self._main_page.task_title_by_title(title).get_attribute("data-testid")
        return data_test_id.replace("task-title-", "")
